app/models/reports/result.rb

```diff
@@ -71,44 +100,57 @@ def last_key(key)
       define_method("#{key}_for") do |period|
         self[key][period]
       end
+
+      define_method("#{key}_for!") do |period|
+        self[key].fetch(period) { raise ArgumentError, "no data found for #{period} for #{key}" }
+      end
     end
 
     # Adjusted registrations are the registrations as of three months ago - we use these for all the percentage
     # calculations to exclude recent registrations.
     def count_adjusted_registrations
-      self.adjusted_registrations = range.each_with_object(Hash.new(0)) do |period, hsh|
+      self.adjusted_registrations = full_data_range.each_with_object(Hash.new(0)) do |period, hsh|
         hsh[period] = cumulative_registrations_for(period.advance(months: -3))
       end
     end
 
+    def count_cumulative_registrations
+      self.cumulative_registrations = full_data_range.each_with_object(Hash.new(0)) { |period, running_totals|
+        previous_registrations = running_totals[period.previous]
+        current_registrations = registrations[period]
+        total = current_registrations + previous_registrations
+        running_totals[period] = total
+      }
+    end
+
     # "Missed visits" is the remaining registerd patients when we subtract out the other three groups.
-    def count_missed_visits
-      self.missed_visits = range.each_with_object({}) { |(period, visit_count), hsh|
+    def calculate_missed_visits(range)
```

Comment:
The fact that some methods like this one take a range to do calculations and others inside Result operate on `full_data_range` not is a sign that this Result object is doing too much and isn't the right abstraction. 

This result object handles the historical results for registrations, controlled/uncontrolled counts, and also missed visits. Some of these are fetched by the ControlRateService but other parts of the data are fetched by other services.  I think there is a follow up PR here to refactor towards a simpler, more focused result object that only cares about a set of data in form of `{period1 => datum1, period2 => datum2, period3 => datum3}`, etc. Then the outer layers will just assemble those results, and the caching parts and ranger filtering parts will become simpler.